### PR TITLE
Initial module pull

### PR DESCRIPTION
**Pull request by Jamie-BitFlight**
## what
* automatically update SSM parameters

## why
* Easily communicate values to other services (E.g. things using `chamber`)

### CODE REVIEW

**Comment by aknysh** [comment]
@Jamie-BitFlight 
very nice module!
we'll work together to improve it, thanks

**Comment by Jamie-BitFlight** [comment]
Thank you. 
Todo draft:
* Make the separation token which is currently `~^~` a variable. Small use case, but easy change.
* fill out the README to cloudposse standard 
* add use case scenarios

File: README.md [hunk 1]
```diff
@@ -1 +1,40 @@
-# terraform-aws-parameter-store
\ No newline at end of file
+# terraform-aws-parameter-store
+
+Terraform module for providing READ and WRITE access to the AWS SSM Parameter Store.
```
Comment (osterman) [inline README.md]: Since "read" and "write" do not appear to be keywords, I think they should be lower-case.

File: example/main.tf [hunk 2]
```diff
@@ -0,0 +1,65 @@
+variable "namespace" {
+  default = "default"
```
Comment (osterman) [inline example/main.tf]: Add `description`

File: example/main.tf [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+variable "tags" {
+  type    = "map"
+  default = {}
```
Comment (osterman) [inline example/main.tf]: add `description`

File: example/main.tf [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+variable "parameter_write" {
+  default = []
+  type    = "list"
```
Comment (osterman) [inline example/main.tf]: Add `description`

File: example/main.tf [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+variable "parameter_read" {
+  type    = "list"
```
Comment (osterman) [inline example/main.tf]: Add `description`

**Comment by osterman** [comment]
> Make the separation token which is currently ~^~ a variable

Agree. Maybe call it `var.delimiter`?

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+*/
+}
+
+variable "tags" {
+  type    = "map"
+  default = {}
+}
+
+variable "kms_arn" {
+  type    = "string"
```
Comment (osterman) [inline main.tf]: Add `description`

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  name = "/${var.namespace}/${var.stage}/${var.name}/database/master_password"
+  # Valid types are String, StringList and SecureString
+  type = "SecureString"
+  value = "${var.master_password}"
+  description = "Database master password"
+  overwrite = false 
+}]
+*/
+}
+
```
Comment (osterman) [inline main.tf]: Move variables to `vars.tf`

File: main.tf [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+variable "tags" {
+  type    = "map"
+  default = {}
```
Comment (osterman) [inline main.tf]: Add `description`

File: main.tf [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  allowed_pattern = "${lookup(var.parameter_write[count.index], "allowed_pattern", "")}"
+  tags            = "${var.tags}"
+}
+
+locals {
+  name_list  = "${compact(concat(split("~^~",join("~^~", aws_ssm_parameter.default.*.name)), split("~^~",join("~^~", data.aws_ssm_parameter.read.*.name))))}"
+  value_list = "${compact(concat(split("~^~",join("~^~", aws_ssm_parameter.default.*.value)), split("~^~",join("~^~", data.aws_ssm_parameter.read.*.value))))}"
+}
+
+output "names" {
```
Comment (osterman) [inline main.tf]: Add `description`

File: main.tf [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  value = "${local.name_list}"
+}
+
+output "values" {
+  value = "${local.value_list}"
+}
+
+output "map" {
```
Comment (osterman) [inline main.tf]: Add `description`

File: main.tf [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+locals {
+  name_list  = "${compact(concat(split("~^~",join("~^~", aws_ssm_parameter.default.*.name)), split("~^~",join("~^~", data.aws_ssm_parameter.read.*.name))))}"
+  value_list = "${compact(concat(split("~^~",join("~^~", aws_ssm_parameter.default.*.value)), split("~^~",join("~^~", data.aws_ssm_parameter.read.*.value))))}"
+}
+
+output "names" {
+  value = "${local.name_list}"
+}
+
+output "values" {
```
Comment (osterman) [inline main.tf]: Add `description`

File: main.tf [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  overwrite       = "${lookup(var.parameter_write[count.index], "overwrite", "false")}"
+  allowed_pattern = "${lookup(var.parameter_write[count.index], "allowed_pattern", "")}"
+  tags            = "${var.tags}"
+}
+
+locals {
+  name_list  = "${compact(concat(split("~^~",join("~^~", aws_ssm_parameter.default.*.name)), split("~^~",join("~^~", data.aws_ssm_parameter.read.*.name))))}"
+  value_list = "${compact(concat(split("~^~",join("~^~", aws_ssm_parameter.default.*.value)), split("~^~",join("~^~", data.aws_ssm_parameter.read.*.value))))}"
+}
+
```
Comment (osterman) [inline main.tf]: Move `outputs` to `outputs.tf` (standard module organization)

File: README.md [hunk 13]
```diff
@@ -1 +1,40 @@
...
+
+* [AWS Details on what values can be used](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-paramstore-su-create.html)
+* [AWS API for PutParameter](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html)
+* [Terraform aws_ssm_parameter resource page](https://www.terraform.io/docs/providers/aws/r/ssm_parameter.html)
+* [Terraform aws_ssm_parameter data page](https://www.terraform.io/docs/providers/aws/d/ssm_parameter.html)
+
+### Simple Write Parameter Example
+```hcl
+module "store_write" {
+  source = "./terraform-aws-parameter-store"
```
Comment (aknysh) [inline README.md]: use `https://github.com/cloudposse/terraform-aws-parameter-store?ref=master`

File: README.md [hunk 14]
```diff
@@ -1 +1,40 @@
...
+  	ManagedBy = "Terraform"
+  }
+}
+```
+This example creates a new `String` parameter called `/production/test/master/company` with the value of `Amazon`
+
+### Simple Read Parameter Example
+```hcl
+module "store_read" {
+  source = "./terraform-aws-parameter-store"
```
Comment (aknysh) [inline README.md]: use `https://github.com/cloudposse/terraform-aws-parameter-store?ref=master`

File: example/main.tf [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+  skip_region_validation      = true
+  skip_credentials_validation = true
+  skip_requesting_account_id  = true
+}
+
+module "kms_key" {
+  source                  = "git::https://github.com/cloudposse/terraform-aws-kms-key.git?ref=master"
+  namespace               = "${var.namespace}"
+  stage                   = "prod"
+  name                    = "key"
```
Comment (aknysh) [inline example/main.tf]: don't use resource types in resource names (this is Cloud Posse's naming convention)
for `name` let's use `app` in all examples

**Comment by osterman** [comment]
In hindsight, wonder if we should rename this module to `terraform-aws-ssm-parameter-store` for preciseness.

**Comment by Jamie-BitFlight** [comment]
Do you guys mind updating the README to match your format when you accept it? Like the credits and stuff?

**Pull request merged**